src/Pirouette/Symbolic/Eval.hs

```diff
@@ -88,18 +98,20 @@ data SymEvalEnv lang = SymEvalEnv
     seeSolvers :: SymEvalSolvers lang
   }
 
--- | A 'SymEval' is equivalent to a function with type:
+data Tag lang = FromBuiltin (BuiltinTerms lang)
+              | FromConstructor Name
+type SymEvalResult lang a = Tree (Tag lang) (SymEvalSt lang) a
+
+-- | Our symbolic evaluation monad.
 --
--- > SymEvalSt lang -> SMT.Solver -> m [(a, SymEvalSt lang)]
+-- It keeps the environment and a connection to the solver in a Reader,
+-- other than that it's the 'TreeT' monad.
 newtype SymEval lang a = SymEval
   { symEval ::
-      ReaderT
-        (SymEvalEnv lang)
-        (StateT (SymEvalSt lang) WeightedList)
-        a
+      ReaderT (SymEvalEnv lang) (TreeT (Tag lang) (SymEvalSt lang) Identity) a
```

Comment:
If we use `TreeT tag Identity`, why not define it as `Tree tag` and explicitely not make it a transformer? Adding any monad other than `Identiy` at the bottom of this stack would be heavily discouraged anyway.

Reply:
That sounds good to me. My initial reaction wa "what if we ever need to go back to `IO`?" But I guess that's for the future...